### PR TITLE
[Rust] Expose DataFragment as pubilc dataset api.

### PR DESCRIPTION
**Pull request by eddyxu**
a `FileFragment` struct modeled after [`pyarrow.dataset.Fragment`](https://arrow.apache.org/docs/python/generated/pyarrow.dataset.Fragment.html)

### CODE REVIEW

File: rust/src/dataset/scanner.rs [hunk 1]
```diff
@@ -93,12 +96,42 @@ impl Scanner {
             offset: None,
             nearest: None,
             with_row_id: false,
+            fragment: None,
```
Comment (changhiskhan) [inline rust/src/dataset/scanner.rs:99]: what's the purpose of a scanner that only does 1 fragment?

File: rust/src/dataset.rs [hunk 2]
```diff
@@ -651,6 +632,18 @@ impl Dataset {
         &self.manifest.schema
     }
 
+    /// Get fragments.
+    ///
+    /// If `filter` is provided, only fragments with the given name will be returned.
+    pub fn get_fragments(&self) -> Vec<FileFragment> {
```
[the remark this answers is not in the document]
Comment (eddyxu) [reply rust/src/dataset.rs:638]: Yea, so i will refactor it later to remove fragments entirely.

File: rust/src/dataset/scanner.rs (hunk 1 above)
Comment (eddyxu) [reply rust/src/dataset/scanner.rs:99]: This is to support `pyarrow.dataset.Fragment.to_batch()`. 
So via exposing `DataFragment` using pyarrow API, users can get a list of Fragments on different machine, and each Fragment can be used as basic unit to read a file, distributedly.

File: rust/src/dataset/scanner.rs [hunk 3]
```diff
@@ -93,12 +96,42 @@ impl Scanner {
...
         }
     }
 
+    pub fn from_fragment(dataset: Arc<Dataset>, fragment: Fragment) -> Self {
```
[the remark this answers is not in the document]
Comment (eddyxu) [reply rust/src/dataset/scanner.rs:103]: `FileFragment::scan()`

File: rust/src/dataset.rs [hunk 4]
```diff
@@ -555,23 +536,23 @@ impl Dataset {
...
-                    &path,
-                    fragment.id,
-                    Some(self.manifest.as_ref()),
+        let fragments = self.get_fragments();
+        let fragment_and_indices = fragments
+            .iter()
+            .map(|f| {
+                (
+                    f,
+                    row_ids_per_fragment.get(&(f.id() as u64)),
```
[the remark this answers is not in the document]
Comment (changhiskhan) [reply rust/src/dataset.rs:545]: oic, so this is not a huge list then. k

**Pull request merged**